lighthouse-core/gather/driver/target-manager.js

```diff
@@ -5,70 +5,125 @@
+   */
+  _findSession(sessionId) {
+    for (const {session, cdpSession} of this._targetIdToTargets.values()) {
+      if (cdpSession.id() === sessionId) return session;
+    }
 
-      // It's not entirely clear when this is necessary, but when the page switches processes on
-      // navigating from about:blank to the `requestedUrl`, resetting `setAutoAttach` has been
-      // necessary in the past.
-      await this._rootCdpSession.send('Target.setAutoAttach', {
-        autoAttach: true,
-        flatten: true,
-        waitForDebuggerOnStart: true,
-      });
-    };
+    throw new Error(`session not found`);
+  }
+
+  /**
+   * Returns the root session.
+   * @return {LH.Gatherer.FRProtocolSession}
+   */
+  rootSession() {
+    const rootSessionId = this._rootCdpSession.id();
+    return this._findSession(rootSessionId);
   }
 
   /**
    * @param {LH.Puppeteer.CDPSession} cdpSession
    */
   async _onSessionAttached(cdpSession) {
-    const session = new ProtocolSession(cdpSession);
+    const newSession = new ProtocolSession(cdpSession);
 
     try {
-      const target = await session.sendCommand('Target.getTargetInfo').catch(() => null);
+      const target = await newSession.sendCommand('Target.getTargetInfo').catch(() => null);
       const targetType = target?.targetInfo?.type;
       const hasValidTargetType = targetType === 'page' || targetType === 'iframe';
+      // TODO: should detach from target in this case?
+      // See pptr: https://github.com/puppeteer/puppeteer/blob/733cbecf487c71483bee8350e37030edb24bc021/src/common/Page.ts#L495-L526
       if (!target || !hasValidTargetType) return;
 
+      // No need to continue if target has already been seen.
       const targetId = target.targetInfo.targetId;
-      session.setTargetInfo(target.targetInfo);
       if (this._targetIdToTargets.has(targetId)) return;
 
+      newSession.setTargetInfo(target.targetInfo);
       const targetName = target.targetInfo.url || target.targetInfo.targetId;
       log.verbose('target-manager', `target ${targetName} attached`);
 
-      const targetWithSession = {target: target.targetInfo, session};
+      const trueProtocolListener = this._getProtocolEventListener(newSession.id());
+      /** @type {(event: unknown) => void} */
+      // @ts-expect-error - pptr currently typed only for single arg emits.
+      const protocolListener = trueProtocolListener;
+      cdpSession.on('*', protocolListener);
+
+      const targetWithSession = {
+        target: target.targetInfo,
+        cdpSession,
+        session: newSession,
+        protocolListener,
+      };
       this._targetIdToTargets.set(targetId, targetWithSession);
-      for (const listener of this._listeners) await listener(targetWithSession);
 
-      await session.sendCommand('Target.setAutoAttach', {
+      // We want to receive information about network requests from iframes, so enable the Network domain.
+      await newSession.sendCommand('Network.enable');
```

Comment:
do you think we should also tweak this to do a promise.all style? https://github.com/GoogleChrome/lighthouse/commit/4bc0b052c187aea0f23ffd704606fa5c0581a7c5
(im still surprised this is fine without it)
also.. i wonder if the `finally` case hits because of a protocol_timeout on one of these?

i know you saw some patterns in pptr/playwright which felt more confident around this stuff. dunno if that's relevant.